src/main/scala/chisel3/util/experimental/decode/decoder.scala

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util.experimental.decode
+
+import chisel3._
+import chisel3.experimental.{ChiselAnnotation, annotate}
+import chisel3.util.{BitPat, pla}
+import chisel3.util.experimental.getAnnotations
+import firrtl.annotations.Annotation
+import logger.LazyLogging
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, Future}
+
+object decoder extends LazyLogging {
+  def apply(minimizer: Minimizer, input: UInt, truthTable: TruthTable, timeout: Int = 5): UInt = {
+    val minimizedTable = getAnnotations().collect {
+      case DecodeTableAnnotation(_, in, out) => TruthTable(in) -> TruthTable(out)
+    }.toMap.getOrElse(
+      {
+        logger.warn(s"""Decoder Cache Hit!
+                        |${truthTable.table}
+                        |""".stripMargin)
+        truthTable
+      }, {
+        val startTime = System.nanoTime()
+        val future = Future(minimizer.minimize(truthTable))
+        var now = System.nanoTime()
+        while (!future.isCompleted) {
+          val elapsed = (System.nanoTime() - now) / 1e9
+          now = System.nanoTime()
+          if(elapsed > timeout) logger.error(s"Minimizer has been executed for ${(System.nanoTime() - startTime) / 1e9} seconds.")
+        }
+        val minimizedTable = Await.result(future, Duration.Inf)
+        val totalTime = System.nanoTime() - startTime
+        val totalTimeInSeconds = totalTime / 1e9
+        val info = f"Logic Minimize with $minimizer finished in $totalTimeInSeconds second"
+        logger.warn(info)
+        minimizedTable
+      }
+    )
```

Comment:
@aswaterman 
I use a `timeout = 5` as default parameter to resolve this, if this minimize function is not finished every 5s, user will be warned. This parameter can be configured to higher if trying to minimize a huge decode table.
My point here is QMC is a NP-hard algorithm, to avoid user directly throw a giant table to espresso or QMC, this notice warns user they are doing a crazy thing.

Reply:
@sequencer I am still not OK with this, for three reasons:
- Since Chisel is embedded in a general-purpose programming language, it is already possible for users to write arbitrary code, including their own infinite loops, NP-hard problems, etc.  It is inconsistent for this one case to emit a warning for long-running code, but not the other cases.
- There are already mechanisms available for debugging [seemingly] non-terminating `java` processes, like `kill -3`.
- This change makes the output non-deterministic, which is a problem for some flows.

Please remove the timeout.

Reply:
Sure!